### ISSUE TITLE
Compliance/Penalty - Late submission of a report triggering a penalty calculation

### ISSUE THREAD

**Issue by NicoleGovvy** [bcgov/cas-compliance#76]
#### Description:

If operators are found to have a previously unknown obligation amount after the Nov. 30 obligation payment deadline, they must meet the compliance obligation and pay a penalty.

Acceptance Criteria:

As an operator, given that I have paid my additional unknown obligation after the Nov. 30 due date, when I log on to the [year] compliance summary page, then I can see in the task list I have an additional compliance obligation penalty that is calculated at prime + 3% **compounding on the first day of every month**.

**Penalty calculation rules:**

- The additional obligation must be paid before the total amount of the penalty is known
- The penalty is charged on the outstanding obligation amount between when it was first overdue (Dec. 1) and the date the additional obligation is paid to 0 
- [Spreadsheet with calculations
](https://bcgov.sharepoint.com/:x:/r/teams/00608-DesignersInternal/_layouts/15/doc2.aspx?sourcedoc=%7B1953CFB9-23FC-4318-8D58-5E2F915D2379%7D&file=Automatic%20penalty%20calculation%20(1).xlsx&action=default&mobileredirect=true&ovuser=6fdb5200-3d0d-4a8a-b036-d3685e359adc%2CNicole.1.Smith%40gov.bc.ca&clickparams=eyJBcHBOYW1lIjoiVGVhbXMtRGVza3RvcCIsIkFwcFZlcnNpb24iOiI1MC8yNDEwMjAwMTMxNiIsIkhhc0ZlZGVyYXRlZFVzZXIiOmZhbHNlfQ%3D%3D)

**NOTE:** Implementing this means that an obligation can have TWO penalties, which we have neither accounted for in our designs, nor in our models. We may want to split out the model updates and do those before prod release.
**Scenario (confirmed as correct by Kirk and John on August 29):**
- A supplementary report resulting in an obligation is created March 15th (several months after compliance deadline of Nov 30)
- The new obligation is not paid to 0 until May 15th
- The obligation will now have TWO penalties:
  - The penalty described in this ticket: Prime +3% from Dec 1 -> May 15th
  - The regular penalty for late payment of an obligation: .38% compounding from April 16th until May 15th (Supplementary obligations past the deadline have a 30 day deadline to pay)

NOTE for @dleard : Let’s consider adding a separate `OneToOneField` for penalties on penalties instead of changing the existing `OneToOneField` to a `ForeignKey`. - Decided on ForeignKey Sept 3 2025

#### Development Checklist:

- [ ] Add a calculation function (calculate_late_submission_penalty()) to the PenaltyCalculationService that handles the penalty scenario described above
- [ ] When calculating each day's compounding values, the function should use the interest_rate value from the elicensing_Interest_rate record where the start_date and end_date contain the date of calculation
- [ ] When creating the penalty record, the penalty type is "late submission"
- [ ] Either add a create_late_submission_penalty() function to the PenaltyCalculationService or refactor create_penalty() to know which penalty type to calculate
- [ ] Update the ObligationPaidHandler handle() function to create a late_submission penalty if the report was created after the compliance deadline (be careful of timezones)
- [ ] Update the PenaltyPaidHandler to _not_ change the penalty_status to PAID unless all related penalties have been paid
- [ ] Update PenaltyCalculationService.create_penalty_invoice() by adding an optional fee_description parameter with a default value ("Automatic Overdue Penalty")
- [ ] Create PenaltyCalculationService.get_late_submission_penalty_data() method that checks if late submission penalty data exists and returns it if found
- [ ] Create a route /compliance-report-versions/{version_id}/late-submission-penalty that return penalty data
- [ ] Add tests
- [ ] Update complianceSummaryReviewSchema.tsx by adding an additional section that displays the calculated penalty information
- [ ] Create getLateSubmissionPenaltyData.ts utility function to fetch penalty data from the backend API
- [ ] Update fetchComplianceSummaryReviewPageData.ts  to include the new penalty data fetching method in the Promise.all() call

#### Definition of Ready (Note: If any of these points are not applicable, mark N/A)

- [ ] User story is included
- [ ] User role and type are identified
- [ ] Acceptance criteria are included
- [ ] Wireframes are included (if required)
- [ ] Design / Solution is accepted by Product Owner
- [ ] Dependencies are identified (technical, business, regulatory/policy)
- [ ] Story has been estimated (under 13 pts)

#### Definition of Done (Note: If any of these points are not applicable, mark N/A)

- [ ] Acceptance criteria are tested by the CI pipeline
- [ ] UI meets accessibility requirements
- [ ] Configuration changes are documented, documentation and designs are updated
- [ ] Passes code peer-review
- [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
- [ ] Ticket is ready to be merged to main branch
- [ ] Can be demoed in Sprint Review
- [ ] Bugs or future work cards are identified and created
- [ ] Reviewed and approved by Product Owner

#### Blocked By/Blocking
Was blocked by #74 
Blocking #227
Was Blocked by Attn Product Owner

#### Notes:

-

**Comment by patrickisaac** [comment]
Refinement Notes:
- Confirming with BA during collab. Ticket will be refined more after the session
- This will be post-MVP

**Comment by patrickisaac** [comment]
Refinement Notes:
- We have discussion with CX
- might be de-prioritized for now

**Comment by dleard** [comment]
Adding the attn: Product owner @vesselak . I wonder if we need another penalty "type" here. This is my understanding of this kind of penalty:

A supplementary report is made 40 days after the compliance deadline of Nov 30
That supplementary report increases the obligation by 100tCO2e
The obligation is due 90 days from invoice creation
There is a _late penalty_ for the supplementary report of prime+3% for each day after Nov 30 until the obligation for the supplementary report is paid. So if they pay it 11 days after the invoice was created, we would add a penalty of prime+3% compounded for 51 days
This penalty is different than the other penalty & will need to be identified in our system
If the supplementary obligation is not paid within the 90 days, then we're doubling up on penalties as the prime+3% penalty is still counting _and_ we're starting to accrue the .38% administrative penalty for late obligation

**Comment by vesselak** [comment]
@dleard I agree that this type of penalty is different from the rest. on the calculation spreadsheet it has its own tab due to the complex calculation as you described in your comment. What do we propose to have this as? Supplementary report after the payment deadline penalty? The late supplementary report is a misleading title but we have time to think about this. I moved the ticket back a sprint for now and it can be done after MVP

**Comment by vesselak** [comment]
@dleard CSNR, Kirk and John have reviewed the ticket text and I have incorporate the minor edits to the text. The scenario has been confirmed as correct so I have removed the blocked labels.

**Comment by vesselak** [comment]
**Note from Kirk:** 
There are, in fact, two separate compliance obligations; the compliance obligation and an additional compliance obligation, for which separate penalty structures apply. The additional compliance obligation can also be subject to the bigger penalty if the operator does not meet it by its due date (which is separate from the compliance obligation deadline). This isn’t to correct what was said, but to spell it out as precisely as I can to avoid misunderstanding 😅. The scenario is correct

**Comment by shon-button** [comment]
@nataliepiguliak 

Nice work on the dev checklists!

I have added some additional points for your consideration:

- Since we now have multiple penalty streams with different rules; perhaps, we need to extend the CompliancePenalty model by adding fields so that each penalty record can independently track its start–end window, the exact interest rates and compounding rules used.
- Before coding, I suggest outlining detailed test scenarios to harden the checklist and de-risk edge cases- something like:
```
* cutoff timing: verify penalties trigger correctly based on report dates around the compliance deadline.
* long-term accrual: confirm penalties accrue over time and compound on the expected schedule.
* short-term accrual: ensure penalties starting later accrue correctly day by day.
* concurrent penalties: validate that multiple penalties can run at the same time and totals combine correctly.
* mid-period payment: check that partial payments reduce outstanding amounts for future accruals.
* boundary payment timing: test how compounding behaves when payment happens on a compounding date.
* changing rates: confirm rate updates are applied correctly across different time periods.

```

**Comment by nataliepiguliak** [comment]
Test Scenarios can be found here:
https://bcgov-my.sharepoint.com/:w:/r/personal/nataliia_piguliak_gov_bc_ca/_layouts/15/doc2.aspx?sourcedoc=%7BFB1F45AC-2BAB-47BA-9209-1642A54875D0%7D&file=76%20Late%20Supplementary%20Report%20Penalty%20Test%20Scenarios.docx&action=default&mobileredirect=true&wdOrigin=OUTLOOK-METAOS.FILEBROWSER

**Comment by dleard** [comment]
Looks like the second item in the DC is incorrect, this is the wording from the [penalties regulation](https://www.bclaws.gov.bc.ca/civix/document/id/lc/statreg/248_2015#section2):

> For the purposes of subsection (7) the rate of interest is, during each successive 3-month period beginning on January 1, April 1, July 1 and October 1 in each year, 3% above the prime lending rate of the principal banker to the Province on the 15th day of the month immediately preceding the 3-month period.

So if I'm reading this correctly: the interest rate should be:
From Oct 1st - Dec 31st: Sep 15th interest rate value
From Jan 1st - Mar 31st: Dec 15th interest rate value
From Apr 1st - Jun 30th: Mar 15th interest rate value
From July 1st - Sep 30th: Jun 15th interest rate value

@vesselak do you agree with my interpretation of the regulation wording?

**Comment by vesselak** [comment]
@nataliepiguliak the correction documented by @dleard above has been confirmed by the business (John Nagle and Kirk Hepburn). Please make the change in the PR. Thank you

**Comment by nataliepiguliak** [comment]
thank you @vesselak! working on it

**Comment by BCerki** [comment]
Some related work was started in https://github.com/bcgov/cas-registration/pull/4003/files, feel free to borrow anything useful

**Comment by fviduya** [comment]
I have tested this in the DEV environment, [here is the test document for reference](https://buttoninc-my.sharepoint.com/:w:/g/personal/francesca_button_is/EZjjPtA6AGNBv8FRklat6cEBI_nMpf-YYEQ6iRG3BmCk9w?e=QK8MfK). During testing, we found that the invoice for the late submission penalty is not being created. Putting this back to In Progress cc: @dleard @vesselak

**Comment by dleard** [comment]
I've created a PR to fix the invoice not being created & something else I found during the fix where the logic to call the create_penalty function wasn't working correctly.
https://github.com/bcgov/cas-registration/pull/4027

**Comment by dleard** [comment]
PR has been merged

**Comment by fviduya** [comment]
I have retested this in the DEV environment, [here is the test document for reference](https://buttoninc-my.sharepoint.com/:w:/g/personal/francesca_button_is/EWDWENTg5BFEv587ZPfEOaUBwdoXQUFapMZr1IojAk3Jyg?e=pG1DU8).

I can verify that:
- When a report is submitted **before** the compliance deadline:
   - And the compliance obligation is paid within 30 days:
      - [x] No penalty is generated
      - [x] No new invoice is created
      - [x] The task list for the late submission penalty is not visible
      - [x] The task list for the automatic overdue penalty is not visible in the frontend
   - And the compliance obligation is paid on the due date:
      - [x] No penalty is generated
      - [x] No new invoice is created
      - [x] The task list for the late submission penalty is not visible
      - [x] The task list for the automatic overdue penalty is not visible in the frontend
   - And the compliance obligation is paid after the due date:
      - [x] An automatic overdue penalty is calculated
      - [x] A new invoice is created for the automatic overdue penalty
      - [x] The task list for the late submission penalty is not visible
      - [x] The task list for the automatic overdue penalty is visible in the frontend
- When a report is submitted **during** the compliance deadline:
   - And the compliance obligation is paid within 30 days:
      - [x] No penalty is generated
      - [x] No new invoice is created
      - [x] The task list for the late submission penalty is not visible
      - [x] The task list for the automatic overdue penalty is not visible in the frontend
   - And the compliance obligation is paid on the due date:
      - [x] No penalty is generated
      - [x] No new invoice is created
      - [x] The task list for the late submission penalty is not visible
      - [x] The task list for the automatic overdue penalty is not visible in the frontend
   - And the compliance obligation is paid after the due date:
      - [x] An automatic overdue penalty is calculated
      - [x] A new invoice is created for the automatic overdue penalty
      - [x] The task list for the late submission penalty is not visible
      - [x] The task list for the automatic overdue penalty is visible in the frontend
- When a report is submitted **after** the compliance deadline:
   - And the compliance obligation is paid within 30 days:
      - [x] A late submission penalty is calculated
      - [x] A new invoice is created for the late submission penalty only
      - [x] The task list for the late submission penalty is visible
      - [ ] The task list for the automatic overdue penalty is not visible in the frontend - **NOTE: this is still visible, but no invoice is created nor a penalty is calculated for this. When user clicks on any of the pages, it returns an error page**
   - And the compliance obligation is paid on the due date:
      - [x] A late submission penalty is calculated
      - [x] A new invoice is created for the late submission penalty only
      - [x] The task list for the late submission penalty is visible
      - [ ] The task list for the automatic overdue penalty is not visible in the frontend - **NOTE: this is still visible, but no invoice is created nor a penalty is calculated for this. When user clicks on any of the pages, it returns an error page**
   - And the compliance obligation is paid after the due date:
      - [x] A late submission penalty is calculated
      - [x] An automatic overdue penalty is calculated
      - [x] New invoices are created for the late submission penalty and automatic overdue penalty
      - [x] The task list for the late submission penalty is visible
      - [x] The task list for the automatic overdue penalty is visible in the frontend
   - When an obligation has both the automatic overdue penalty and late submission penalty:
      - [x] When user pays the automatic overdue penalty only, and `run_scheduled_compliance_sync()` is executed, the Penalty Status is still set to `Not Paid`
      - [ ] When the automatic overdue penalty is paid, the Penalty Status in the Review Penalty Summary page is updated to `Paid` after `run_scheduled_compliance_sync()` is executed - **NOTE: Even though the payment has already been posted for this penalty, and the outstanding balance is at 0, the Penalty Status is still set to `Not Paid`**
      - [x] When all penalties have been paid, and `run_scheduled_compliance_sync()` is executed, the Penalty Status is updated to `Paid`
- [x] When a report is submitted before or during the compliance period, the invoice due date is set to November 30
- [x] When a report is submitted after the compliance period, the invoice due date is set to 30 days after the compliance obligation is created 
- [x] When a penalty exists for the compliance report version but the compliance obligation has already been paid, and `run_scheduled_compliance_sync()` is executed, the Compliance Status is updated to `Obligation - met` 

Hi @vesselak, the only issue found during testing were more about the frontend, since we still have #331 coming in I was planning on testing it first and if the issues found here are not resolved in that ticket, then we can make a new one to tackle outstanding issues. Please let me know if that works for you. I'm moving this to Pending PO for the review of test. Thank you!

**Comment by vesselak** [comment]
Review testing documentation and accepting ticket. Amazing work @nataliepiguliak @dleard and @fviduya 

I agree with the approach to create the tickets after testing #331 as the deficiencies are not blockers and/or are likely addressed in the UI ticket.

**Comment by TammieV-BCG** [comment]
UAT by Business area. 2 Defect found. [Here is the report for your reference](https://bcgov.sharepoint.com/:w:/r/teams/00608-ScrumTeam/Shared%20Documents/Scrum%20Team/OBPS/UAT/Defect%20reports/Defect%20Report-76.docx?d=w718811d421364ad2885678c5c0c42968&csf=1&web=1&e=Ui8gCL)

**Comment by vesselak** [comment]
Based on UAT findings a fix will be processed today. Kirk Hepburn provided the following changes that will be included in the fix:
**Supplementary:**

- small penalty calculated from compliance_deadline to payment date
- large penalty applied **if not paid within 30 days of submission**

**Initial Report:**

- ~small penalty calculated from compliance_deadline to payment date~
- large penalty calculated from compliance_deadline to payment date

cc @dleard @fviduya @TammieV-BCG

**Comment by TammieV-BCG** [comment]
UAT by Business Area. Works as expected

**Comment by vesselak** [comment]
Released to production with v4.5.0
